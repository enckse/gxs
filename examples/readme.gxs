# comments start with '#' but can NOT be within a '{' and '}' block
palette => {
    x => red
    y => #333333
    z => NONE
}
mode => {
    xs
}
pattern => {
    xyzyz
    zyxyz
    zzxzz
}
action => {
    commit
}
# reuse the palette
mode => {
    te
}
# patterns must always be redefined
pattern => {
    xxxxx
    zzzzz
    xxxxx
}
action => {
    commit
}
