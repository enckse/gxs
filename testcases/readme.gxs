# comments start with '#' but can NOT be within a '{' and '}' block
palette => {
    x => red
    y => #333333
    z => NONE
}
mode => {
    xstitch
}
pattern => {
    xyzyz
    zyxyz
    zzxzz
}
action => {
    commit
}
# reuse the palette
mode => {
    topedge
}
# patterns must always be redefined
pattern => {
    xxxxx
    zzzzz
    xxxxx
}
action => {
    commit
}
