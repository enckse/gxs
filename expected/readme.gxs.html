<!doctype html>
<html lang="en">
<head>
<meta charset="UTF-8">
<style>
.container {
  background: white;
  display: inline-block;
  border: 1px dotted black;
}
.grid {
  display: grid;
  grid-template-columns: repeat(6, 10px);
  grid-template-rows: repeat(6, 10px);
  grid-gap: 1px;
}
.cell {
  justify-content: center;
  align-items: center;
  display: flex;
  font-family: Arial;
  font-size: 4pt;
  font-weight: bold;
  background: white;
}
.main {
  margin-left: 10px;
  padding: 0px 10px;
}
</style>
<script>

function make_grid_id(x, y) {
    return ("000" + x).slice(- 3 ) + "x" + ("000" + y).slice(- 3 );
}

document.onreadystatechange = function () {
    if (document.readyState == "complete") {
        try {
            build();
        } catch (error) {
            alert(error);
        }
    }
}

function expand_range(token) {
    var parts = token.split("-");
    var results = [];
    var first = parseInt(parts[0]);
    if (parts.length == 1) {
        results.push(first);
    } else {
        if (parts.length > 2) {
            throw "invalid split, range must be M-N only: " + token;
        }
        second = parseInt(parts[1]);
        if (second <= first) {
            throw "invalid range, N > M: " + token;
        }
        var x = first;
        while (x <= second) {
            results.push(x);
            x += 1;
        }
    }
    console.log(results);
    return results;
}

function build() {
    var obj = [{"cells":["1x1","3x2","3x3"],"mode":"xs","color":"rgb(199, 43, 59)"},{"cells":["2x1","4x1","2x2","4x2"],"mode":"xs","color":"#333333"},{"cells":["1x1","2x1","3x1","4x1","5x1","1x3","2x3","3x3","4x3","5x3"],"mode":"te","color":"rgb(199, 43, 59)"}];
    var width =  6 ;
    var height =  6 ;
    var history = {};
    var split_char = "x";
    for (var i = 0; i < obj.length; i++) {
        var entry = obj[i];
        var color = entry["color"];
        var mode = entry["mode"];
        var cells = entry["cells"];
        var items = []
        for (var j = 0; j < cells.length; j++) {
            var cell = cells[j];
            var parts = cell.split(split_char);
            var x = expand_range(parts[1]);
            var y = expand_range(parts[0]);
            var range_x = [];
            var range_y = [];
            for (var k = 0; k < x.length; k++) {
                for (var l = 0; l < y.length; l++) {
                    console.log(x[k].toString() + split_char + y[l].toString());
                    items.push(x[k].toString() + split_char + y[l].toString());
                }
            }
        }
        for (var j = 0; j < items.length; j++) {
            var cell = items[j];
            var parts = cell.split(split_char);
            var raw_x = parts[0];
            var raw_y = parts[1];
            var x = parseInt(raw_x);
            var y = parseInt(raw_y);
            if (x <= 0 || y <= 0) {
                throw "found <= 0 cell (must be > 0): " + cell;
            }
            if (x >= width || y >= height) {
                throw "found cell outside height/width: " + cell;
            }
            var pad = make_grid_id(raw_x, raw_y);
            var element = document.getElementById(pad);
            var style = "";
            if (pad in history) {
                style = history[pad];
            }
            if (mode === "te") {
                style += " border-top-style: solid; border-top-color: ";
            } else if (mode == "be") {
                style += " border-bottom-style: solid; border-bottom-color: ";
            } else if (mode == "re") {
                style += " border-right-style: solid; border-right-color: ";
            } else if (mode == "le") {
                style += " border-left-style: solid; border-left-color: ";
            } else if (mode == "xs") {
                style += " background-color: ";
            } else {
                throw "unknown mode";
            }
            style += color + ";";
            element.setAttribute("style", style);
            history[pad] = style;
        }
    }
}
</script>
    </head>
    <body>
        <div class="main">
  <div class="container">
  <div class="grid" id="grid">
      <div class="cell" id="000x000"></div>
      <div class="cell" id="000x001">1</div>
      <div class="cell" id="000x002">2</div>
      <div class="cell" id="000x003">3</div>
      <div class="cell" id="000x004">4</div>
      <div class="cell" id="000x005">5</div>
      <div class="cell" id="001x000">1</div>
      <div class="cell" id="001x001"></div>
      <div class="cell" id="001x002"></div>
      <div class="cell" id="001x003"></div>
      <div class="cell" id="001x004"></div>
      <div class="cell" id="001x005"></div>
      <div class="cell" id="002x000">2</div>
      <div class="cell" id="002x001"></div>
      <div class="cell" id="002x002"></div>
      <div class="cell" id="002x003"></div>
      <div class="cell" id="002x004"></div>
      <div class="cell" id="002x005"></div>
      <div class="cell" id="003x000">3</div>
      <div class="cell" id="003x001"></div>
      <div class="cell" id="003x002"></div>
      <div class="cell" id="003x003"></div>
      <div class="cell" id="003x004"></div>
      <div class="cell" id="003x005"></div>
      <div class="cell" id="004x000">4</div>
      <div class="cell" id="004x001"></div>
      <div class="cell" id="004x002"></div>
      <div class="cell" id="004x003"></div>
      <div class="cell" id="004x004"></div>
      <div class="cell" id="004x005"></div>
      <div class="cell" id="005x000">5</div>
      <div class="cell" id="005x001"></div>
      <div class="cell" id="005x002"></div>
      <div class="cell" id="005x003"></div>
      <div class="cell" id="005x004"></div>
      <div class="cell" id="005x005"></div>
  </div>
</div>
        </div>
    </body>
</html>
